miden-crypto/src/merkle/smt/forest/mod.rs

```diff
@@ -0,0 +1,247 @@
+
+    /// Creates an empty `SmtForest` instance.
+    pub fn new() -> SmtForest {
+        let roots = BTreeSet::new();
+        let store = SmtStore::new();
+        let leaves = Map::new();
+
+        SmtForest { roots, store, leaves }
+    }
+
+    // DATA EXTRACTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an opening for the specified key in the SMT with the specified root.
+    ///
+    /// Returns an error if an SMT with this root is not in the forest, or if the forest does
+    /// not have sufficient data to provide an opening for the specified key.
+    pub fn open(&self, root: Word, key: Word) -> Result<SmtProof, MerkleError> {
+        if !self.contains_root(root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        let leaf_index = NodeIndex::from(LeafIndex::from(key));
+
+        let proof = self.store.get_path(root, leaf_index)?;
+        let path = proof.path.try_into()?;
+        let leaf = proof.value;
+
+        let Some(leaf) = self.leaves.get(&leaf).cloned() else {
+            return Err(MerkleError::UntrackedKey(key));
+        };
+
+        SmtProof::new(path, leaf).map_err(|error| match error {
+            SmtProofError::InvalidMerklePathLength(depth) => MerkleError::InvalidPathLength(depth),
+        })
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts the specified key-value pair into an SMT with the specified root. This will also
+    /// add a new root to the forest.
+    ///
+    /// Returns an error if an SMT with the specified root is not in the forest, these is not
+    /// enough data in the forest to perform the insert, or if the insert would create a leaf
+    /// with too many entries.
+    pub fn insert(&mut self, root: Word, key: Word, value: Word) -> Result<Word, MerkleError> {
+        self.batch_insert(root, vec![(key, value)])
+    }
+
+    /// Inserts the specified key-value pairs into an SMT with the specified root. This will also
+    /// add a single new root to the forest for the entire batch of inserts.
+    ///
+    /// Returns an error if an SMT with the specified root is not in the forest, these is not
+    /// enough data in the forest to perform the insert, or if the insert would create a leaf
+    /// with too many entries.
+    pub fn batch_insert(
+        &mut self,
+        root: Word,
+        entries: impl IntoIterator<Item = (Word, Word)> + Clone,
+    ) -> Result<Word, MerkleError> {
+        if !self.contains_root(root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        // Find all affected leaf indices
+        let indices = entries
+            .clone()
+            .into_iter()
+            .map(|(key, _)| LeafIndex::from(key))
+            .collect::<BTreeSet<_>>();
```

Comment:
nit: Couldn't this just be

```suggestion
        let indices = entries
            .iter()
            .map(|(key, _)| LeafIndex::from(*key))
            .collect::<BTreeSet<_>>();
```

and then I think you could drop the `Clone` bound. Although the difference is probably minimal